### ISSUE TITLE
Unexpected warning using parameter shift gradient with NumPy backend

### ISSUE THREAD

**Issue by liwt31** [tencent-quantum-lab/tensorcircuit#92]
## Issue Description

Unexpected warning using parameter shift gradient with NumPy backend.
The problem is at
https://github.com/tencent-quantum-lab/tensorcircuit/blob/df6679561e71cbc3eec1d3f5f2c731b008591a8f/tensorcircuit/experimental.py#L247

The problem is probably also present in `parameter_shift_grad_v2`.

## How to Reproduce

```python
import numpy as np

import tensorcircuit as tc
from tensorcircuit import experimental as E

K = tc.set_backend("numpy")


def f1(param):
    c = tc.Circuit(1)
    c.rz(0, theta=param[0])
    return K.real(c.expectation_ps(y=[0]))

E.parameter_shift_grad(f1)(np.array([1]))
```
This is more generally an inconsistency between backends. JAX and NumPy will issue a warning when casting from complex to float while tensorflow will not. 
```python
import tensorcircuit as tc

K = tc.set_backend("jax")
a = K.ones(1)
K.cast(a, float)
```
NumPy and JAX use `astype` for `casting` while tensorflow uses plain `cast`. Since the function name is `cast`, it makes more sense to adopt the tensorflow behavior and suppress the warning.

### Error Output

```
ComplexWarning: Casting complex values to real discards the imaginary part
```

## Environment Context

commit df6679561e71cbc3eec1d3f5f2c731b008591a8f

**Comment by refraction-ray** [comment]
Yes, there are always many warnings on implicit type convention when using jax, in general no numerical bug. Any specific implementation idea to screen this type of warning in tc?

**Comment by liwt31** [comment]
Two ideas:
1. The warnings can be disabled globally with
```python
import warnings
from numpy import ComplexWarning
warnings.filterwarnings("ignore", category=ComplexWarning)
jnp.ones([2], dtype=jnp.complex64).astype(jnp.float32)
```
It is also possible the suppress the warning temporarily instead of globally according to the [documentation](https://docs.python.org/3/library/warnings.html#temporarily-suppressing-warnings).
2. In the `backend.cast` method, check complex to float conversion and raise an exception when the imaginary part is not zero or just directly take the real part. 

> And this warning is from implicit dtype conversion, instead of astype?

I think so

**Comment by refraction-ray** [comment]
Thanks for your suggestion, will take the following approach within cast method implementation

```python
with warnings.catch_warnings() as w:
    warnings.simplefilter("ignore", np.ComplexWarning)
    jnp.ones([2], dtype=jnp.complex64).astype(jnp.float32)
```

Note before python 3.11, ``catch_warnings`` doesn't accept `category` parameter
